device/transport/mqtt/devdoc/mqtt_twin_client_requirements.md

```diff
@@ -0,0 +1,95 @@
+# MqttTwinClient Requirements
+
+## Overview
+Object used to publish, subscribe and receive twin messages.
+
+## Responsibilities
+- subscribe to the necessary topics to achieve twin operations
+- format MQTT messages to send twin requests and reported properties updates
+- get the full twin using the appropriate topics and request/response mechanism
+- receive desired properties updates and signal them to the Mqtt client.
+
+## Non-responsibilities
+- connecting/disconnecting the Mqtt connection
+- any other feature that is not twin-related
+- retries and timeouts.
+
+## Public API
+```typescript
+class MqttTwinClient extends EventEmitter {
+  constructor(client: MqttBase);
```

Comment:
>MqttBase [](start = 22, length = 8)

Architectural question: If this object accepts an MqttBase object, that has a weak implication that the MqttTwinClient object has some sort of independence from the Mqtt object.  I find that interesting, but this is adding another level of nested FSM and that's starting to get confusing to me. 
I don't think I'm explaining this too well, but this object has an FSM, but that FSM makes assumptions about the state of the base transport.  e.g. this FSM assumes that the base is connected, do device/mqtt creates the base object, and is responsible for connecting it, but this object uses it.
Still sorting this out....smells bad though.

Reply:
The more I think about it, I'm not sure why twin_client is separated out.  Are you doing it to keep your .ts file smaller?  If so, this feels like a partial class.  Encapsulation just feels strange here.  

---
In reply to: [170618901](https://github.com/Azure/azure-iot-sdk-node/pull/236#discussion_r170618901) [](ancestors = 170618901)

Reply:
yes and yes. see my other comment. there's definitely a smell to this and I'm open to suggestions.

---
In reply to: [170704664](https://github.com/Azure/azure-iot-sdk-node/pull/236#discussion_r170704664) [](ancestors = 170704664,170618901)